### PR TITLE
[feature](Nereids): add or expansion in CBO

### PR DESCRIPTION
**Pull request by keanji-x**
## Proposed changes

add or expansion in CBD. The rule is :
```
  https://blogs.oracle.com/optimizer/post/optimizer-transformations-or-expansion
  NLJ (cond1 or cond2)              UnionAll
                        =>         /        \
                         HJ(cond1)   HJ(cond2 and !cond1)
```
TODO: 
* support expanding complex conditions  

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.62 seconds
 stream load tsv:          514 seconds loaded 74807831229 Bytes, about 138 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          32 seconds loaded 861443392 Bytes, about 25 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17161800259 Bytes

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.75 seconds
 stream load tsv:          509 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162198804 Bytes

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.02 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17162261791 Bytes

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.51 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.1 seconds inserted 10000000 Rows, about 343K ops/s
 storage size: 17162435102 Bytes

**Comment by XieJiann** [comment]
run buildall

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.45 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162063206 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.28 seconds
 stream load tsv:          511 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         31 seconds loaded 2358488459 Bytes, about 72 MB/s
 stream load orc:          71 seconds loaded 1101869774 Bytes, about 14 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162466834 Bytes

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+            LogicalJoin<? extends Plan, ? extends Plan> newJoin = join.withJoinConjuncts(hashCond, otherCond)
+                    .withChildren(Lists.newArrayList(left, right));
+            joins.add(newJoin);
+        }
+        return joins;
+    }
+
+    // For disjunctions: a or b or c
+    // we can get: (b and !a) union all (c and !a and !b)
+    private Pair<List<Expression>, List<Expression>> extractCondition(int index, List<Expression> equal,
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java:146]: explain the result Pair

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                    LogicalUnion union = new LogicalUnion(Qualifier.ALL, new ArrayList<>(join.getOutput()),
+                            ImmutableList.of(),
+                            false, joins);
+                    LogicalCTEAnchor<? extends Plan, ? extends Plan> intermediateAnchor = new LogicalCTEAnchor<>(
+                            rightProducer.getCteId(), rightProducer, union);
+                    return new LogicalCTEAnchor<Plan, Plan>(leftProducer.getCteId(), leftProducer, intermediateAnchor);
+                }).toRule(RuleType.OR_EXPANSION);
+    }
+
+    private Pair<List<Expression>, List<Expression>> expandExpr(Expression otherExpr,
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java:106]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+    private List<Plan> expandJoin(StatementContext ctx, List<Expression> disjunctions, List<Expression> otherConditions,
+            LogicalJoin<? extends Plan, ? extends Plan> join, LogicalCTEProducer<? extends Plan> leftProducer,
+            LogicalCTEProducer<? extends Plan> rightProducer) {
+        List<Expression> notExprs = disjunctions.stream().map(Not::new).collect(Collectors.toList());
+        List<Plan> joins = Lists.newArrayList();
+
+        for (int i = 0; i < disjunctions.size(); i++) {
+            // construct other condition and hash condition
+            Pair<List<Expression>, List<Expression>> pair = extractCondition(i, disjunctions, notExprs);
```
[the remark this answers is not in the document]
Comment (XieJiann) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/exploration/OrExpansion.java]: done

**Comment by XieJiann** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 48.23 seconds
 stream load tsv:          512 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17161997341 Bytes

**Review by morrySnow — APPROVED** [review]

**Review by jackwener — APPROVED** [review]

**Pull request merged**